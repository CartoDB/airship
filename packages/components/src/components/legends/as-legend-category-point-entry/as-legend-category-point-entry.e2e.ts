import { E2EPage, newE2EPage } from '@stencil/core/testing';

describe('as-legend-category-point-entry', async () => {
  describe('Rendering', async () => {
    let page: E2EPage;

    it('basic points', async () => {
      page = await newE2EPage({ html: `
        <as-legend-category-point-entry>
        </as-legend-category-point-entry>
      ` });

      const element = await page.find('as-legend-category-point-entry');
      element.setProperty('label', 'Hello there');
      element.setProperty('color', '#FABADA');
      element.setProperty('strokeColor', '#C0C1D0');
      element.setProperty('strokeStyle', 'dashed');
      element.setProperty('width', '16');
      await page.waitForChanges();
      expect(element.outerHTML).toMatchSnapshot();
    });

    it('markers', async () => {
      page = await newE2EPage({ html: `
        <as-legend-category-point-entry>
        </as-legend-category-point-entry>
      ` });

      const element = await page.find('as-legend-category-point-entry');
      element.setProperty('label', 'Hello there');
      element.setProperty('color', '#FABADA');
      element.setProperty('strokeColor', '#C0C1D0');
      element.setProperty('strokeStyle', 'dashed');
      element.setProperty('marker', 'wadus.svg');
      element.setProperty('width', '16');
      await page.waitForChanges();
      expect(element.outerHTML).toMatchSnapshot();
    });
  });
});
