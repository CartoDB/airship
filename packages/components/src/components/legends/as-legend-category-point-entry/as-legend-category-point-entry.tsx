import { Component, Prop } from '@stencil/core';
import { borderStyleCounts } from '../../../utils/border-style-counts';

// This component ignores the strokeWidth property, and always paints
// a 1px border.
const FAKE_BORDER_SIZE = 1;

const DEFAULT_WIDTH = 16;
// Required for background masks to work properly
const SYMBOL_OFFSET = 2;

@Component({
  shadow: false,
  styleUrl: './as-legend-category-point-entry.scss',
  tag: 'as-legend-category-point-entry',
})
export class LegendCategoryPointEntry {
  @Prop() public label: string;
  @Prop() public color: string;
  @Prop() public strokeColor: string;
  @Prop() public strokeStyle: string;
  @Prop() public marker: string;
  @Prop() public width: number = DEFAULT_WIDTH;

  public render() {
    const classes = {
      'as-legend-category-point-entry--point': !this.marker,
      'as-legend-category-point-entry--symbol': true
    };

    return (
      <div class='as-legend-category-point-entry--wrapper'>
        <div class='as-legend-category--figure--wrapper'>
          <div class={classes} style={this.getStyle()}></div>
        </div>
        <span class='as-legend-category--label'>{this.label}</span>
      </div>
    );
  }

  private getStyle() {
    return {
      backgroundColor: this.color,
      border: `${FAKE_BORDER_SIZE}px ${this.strokeStyle || 'solid'} ${this.strokeColor}`,
      ...this.getWidth(),
      ...this.getMask()
    };
  }

  private getWidth() {
    if (!this.width) {
      return {};
    }

    let sizeOffset = borderStyleCounts(this.strokeStyle)
      ? FAKE_BORDER_SIZE * 2
      : 0;

    if (this.marker) {
      sizeOffset += SYMBOL_OFFSET;
    }

    return {
      height: `${this.width + sizeOffset}px`,
      width: `${this.width + sizeOffset}px`
    };
  }

  private getMask() {
    if (!this.marker) {
      return {};
    }

    return {
      '-webkit-mask-image': `url(${this.marker})`,
      '-webkit-mask-position': 'center',
      '-webkit-mask-repeat': 'no-repeat',
      '-webkit-mask-size': `${this.width}px`,
      'maskImage': `url(${this.marker})`,
      'maskPosition': 'center',
      'maskRepeat': 'no-repeat',
      'maskSize': `${this.width}px`
    };
  }
}
