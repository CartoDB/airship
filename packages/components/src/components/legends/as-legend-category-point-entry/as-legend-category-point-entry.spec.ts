import { newSpecPage, SpecPage } from '@stencil/core/dist/testing';
import { LegendCategoryPointEntry } from './as-legend-category-point-entry';

describe('as-legend-category-point-entry', () => {
  describe('Rendering', () => {
    let page: SpecPage;

    it('basic points', async () => {
      page = await newSpecPage({
        components: [LegendCategoryPointEntry],
        html: `<as-legend-category-point-entry></as-legend-category-point-entry>`
      });

      const element = page.root;
      element.label = 'Hello there';
      element.color = '#FABADA';
      element.strokeColor = '#C0C1D0';
      element.strokeStyle = 'dashed';
      element.width = '16';
      await page.waitForChanges();
      expect(element.outerHTML).toMatchSnapshot();
    });

    it('markers', async () => {
      page = await newSpecPage({
        components: [LegendCategoryPointEntry],
        html: `<as-legend-category-point-entry></as-legend-category-point-entry>`
      });

      const element = page.root;
      element.label = 'Hello there';
      element.color = '#FABADA';
      element.strokeColor = '#C0C1D0';
      element.strokeStyle = 'dashed';
      element.marker = 'wadus.svg';
      element.width = '16';
      await page.waitForChanges();
      expect(element.outerHTML).toMatchSnapshot();
    });
  });
});
