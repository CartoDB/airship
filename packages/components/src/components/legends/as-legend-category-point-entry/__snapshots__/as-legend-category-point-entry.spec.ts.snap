// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`as-legend-category-point-entry Rendering basic points 1`] = `"<as-legend-category-point-entry><div class=\\"as-legend-category-point-entry--wrapper\\"><div class=\\"as-legend-category--figure--wrapper\\"><div class=\\"as-legend-category-point-entry--point as-legend-category-point-entry--symbol\\" style=\\"border: 1px dashed #C0C1D0; height: 18px; width: 18px; background-color: #FABADA;\\"></div></div><span class=\\"as-legend-category--label\\">Hello there</span></div></as-legend-category-point-entry>"`;

exports[`as-legend-category-point-entry Rendering markers 1`] = `"<as-legend-category-point-entry><div class=\\"as-legend-category-point-entry--wrapper\\"><div class=\\"as-legend-category--figure--wrapper\\"><div class=\\"as-legend-category-point-entry--symbol\\" style=\\"border: 1px dashed #C0C1D0; height: 20px; width: 20px; background-color: #FABADA; -webkit-mask-image: url(wadus.svg); -webkit-mask-position: center; -webkit-mask-repeat: no-repeat; -webkit-mask-size: 16px; mask-image: url(wadus.svg); mask-position: center; mask-repeat: no-repeat; mask-size: 16px;\\"></div></div><span class=\\"as-legend-category--label\\">Hello there</span></div></as-legend-category-point-entry>"`;
